superset/assets/visualizations/rose.jsx

```diff
@@ -0,0 +1,539 @@
+          startAngle, endAngle, innerRadius, name, outerRadius: outerRadius + gro,
+        };
+        arcSt.push[arcId] = {
+          startAngle, endAngle, innerRadius: innerRadius + gro, outerRadius: outerRadius + gro,
+        };
+        arcSt.pieStart[arcId] = {
+          startAngle, endAngle, innerRadius: mini * maxRadius, outerRadius: maxRadius,
+        };
+        arcSt.mini[arcId] = {
+          startAngle, endAngle, innerRadius: innerRadius * mini, outerRadius: outerRadius * mini,
+        };
+        arcId++;
+        innerRadius = outerRadius;
+      }
+      const labelArc = Object.assign({}, arcSt.data[i * numGroups]);
+      labelArc.outerRadius = maxRadius + 20;
+      labelArc.innerRadius = maxRadius + 15;
+      arcSt.labels.push(labelArc);
+      for (const v of adatum[t].concat().sort(sortValues)) {
+        const val = state.disabled[v.id % numGroups] ? 0 : v.value;
+        pieEndAngle = G * val + pieStartAngle;
+        arcSt.pie[v.id] = {
+          startAngle: pieStartAngle,
+          endAngle: pieEndAngle,
+          innerRadius: maxRadius * mini,
+          outerRadius: maxRadius,
+          percent: v.value / sums[i],
+        };
+        arcSt.pieOver[v.id] = {
+          startAngle: pieStartAngle,
+          endAngle: pieEndAngle,
+          innerRadius: maxRadius * mini,
+          outerRadius: maxRadius + gro,
+        };
+        pieStartAngle = pieEndAngle;
+      }
+    }
+    arcSt.groupLabels = arcSt.data.slice(0, numGroups);
+    return arcSt;
+  }
+
+  let arcSt = computeArcStates(datum);
+
+  function tween(target, resFunc) {
+    return function (d) {
+      const interpolate = d3.interpolate(copyArc(d), copyArc(target));
+      return t => resFunc(Object.assign(d, interpolate(t)));
+    };
+  }
+
+  function arcTween(target) {
+    return tween(target, d => arc(d));
+  }
+
+  function translateTween(target) {
+    return tween(target, d => `translate(${arc.centroid(d)})`);
+  }
+
+  // Grab the ID range of segments stand between
+  // this segment and the edge of the circle
+  const segmentsToEdgeCache = {};
+  function getSegmentsToEdge(arcId) {
+    if (segmentsToEdgeCache[arcId]) {
+      return segmentsToEdgeCache[arcId];
+    }
+    const timeIndex = Math.floor(arcId / numGroups);
+    segmentsToEdgeCache[arcId] = [arcId + 1, numGroups * (timeIndex + 1) - 1];
+    return segmentsToEdgeCache[arcId];
+  }
+
+  // Get the IDs of all segments in a timeIndex
+  const segmentsInTimeCache = {};
+  function getSegmentsInTime(arcId) {
+    if (segmentsInTimeCache[arcId]) {
+      return segmentsInTimeCache[arcId];
+    }
+    const timeIndex = Math.floor(arcId / numGroups);
+    segmentsInTimeCache[arcId] = [timeIndex * numGroups, (timeIndex + 1) * numGroups - 1];
+    return segmentsInTimeCache[arcId];
+  }
+
+  let clickId = -1;
+  let inTransition = false;
+  const ae = roseWrap
+    .selectAll('g')
+    .data(JSON.parse(JSON.stringify(arcSt.data))) // deep copy data state
```

Comment:
why does this need to be deep copied?

Reply:
The object passed to `.data()` becomes the 'state' of the arcs and is modified when they transition. `arcSt.data` needs to be copied so that the arcs can transition back to their original position.